Добавляем в запущенный питон всякие дополнительные модули, которые нам понадобятся

In [2]:
import tensorflow as tf # нейросетка
import keras # модуль, который упрощает работу с нейросеткой
import os
import numpy as np # модуль для особого формата чисел
import pandas as pd # модуль для таблиц, чтобы удобнее было смотреть на цифирки
from keras.models import Sequential
from keras.layers import Dense

Для примера создадим датасет себе сами,
чтобы нам было понятнее что там внутри
как что с чем связано и что мы захотим от нейросетки увидеть,
когда она заработает.

В датасете будет три параметра: один искомый/интересующий, 
который мы ждём от нейросетки на выходе - это y. И два будем подавать на вход (это x1 и x2) - то есть от этих параметров должен зависеть результат (y)

Сделаем датасет так, что y (интересуемый) должен быть почти на единицу меньше, чем два входных параметра.
В результате получим таблицу с тремя столбцами (y, x1, x2) и тысячей строк

In [3]:
x = list()
y = list()

for i in range(1000):  # цикл 1000 раз
    x.append([i+1.2, i+2.1])  # эта переменная хранит в себе два столбца - x1 и x2
    y.append(i)

Создадим удобную таблицу и посмотрим как выглядит её начало:

In [4]:
df = pd.DataFrame(x, columns=['x1', 'x2'])
df['y'] = y
df = df[['y', 'x1', 'x2']]
df.head()

y   x1   x2
0  0  1.2  2.1
1  1  2.2  3.1
2  2  3.2  4.1
3  3  4.2  5.1
4  4  5.2  6.1

Теперь нужно разбить нашу таблицу на две группы. Одна группа это данные, по которой модель учится (train). Вторая группа (test) это данные, по которой мы будем проверять работу модели - эти данные не должны попасть в корпус обучения. Так мы посмотрим как модель работает на данных, которые не видела.

Пусть, первые 700 строк таблицы это будут данные для обучения, а последние триста это данные для теста. Без разницы.

In [5]:
x_train = np.array(x[:700])
x_test = np.array(x[700:])
y_train = np.array(y[:700])
y_test = np.array(y[700:])

Посмотрим на тестовую выборку:

In [6]:
df = pd.DataFrame(x_test, columns=['x1_test', 'x2_test'])
df['y_test'] = y_test
df = df[['y_test', 'x1_test', 'x2_test']]
df.head()

y_test  x1_test  x2_test
0     700    701.2    702.1
1     701    702.2    703.1
2     702    703.2    704.1
3     703    704.2    705.1
4     704    705.2    706.1

Сделаем самую простую из возможных нейросеток для данной задачи: два нейрона принимают два параметра (x1 и x2), один нейрон выдаёт наш искомый y

In [16]:
model = Sequential()

# первый слой - два нейрона принимают два параметра
model.add(Dense(2, input_dim=2, kernel_initializer='normal', activation='relu'))

# второй слой - один нейрон
model.add(Dense(1, kernel_initializer='normal'))

model.compile(loss='mean_squared_error', optimizer='adam')

Обучаем (тренируем) модель на тренировочном датасете, который состоит первых 700 строк. Переменная x_train включает в себя два параметра, если что (x1 и x2)

In [17]:
model.fit(x_train, y_train, epochs=100)

Epoch 1/100
 32/700 [>.............................] - ETA: 3s - loss: 124504.8828700/700 [==============================] - 0s 273us/step - loss: 163930.1086
Epoch 2/100
700/700 [==============================] - 0s 36us/step - loss: 163145.9828
Epoch 3/100
700/700 [==============================] - 0s 47us/step - loss: 162945.6529
Epoch 4/100
700/700 [==============================] - 0s 38us/step - loss: 162930.2849
Epoch 5/100
700/700 [==============================] - 0s 52us/step - loss: 162914.9480
Epoch 6/100
700/700 [==============================] - 0s 32us/step - loss: 162899.6125
Epoch 

Чтобы проверить как модель обучилась, нужно подать ей на вход тестовые x1 и x2, которые ей не известны (последние 300 строк таблицы) и посмотреть какой "y" она нам предскажет. Такой ли, какой мы от неё ждём

In [18]:
y_predicted = model.predict(x_test)

Сделаем таблицу:

In [19]:
df = pd.DataFrame(x_test, columns=['x1_test', 'x2_test'])
df['y_predicted'] = y_predicted
df = df[['y_predicted', 'x1_test', 'x2_test']]
df.head()

y_predicted  x1_test  x2_test
0     2.186424    701.2    702.1
1     2.186424    702.2    703.1
2     2.186424    703.2    704.1
3     2.186424    704.2    705.1
4     2.186424    705.2    706.1

Получилась какая-то хуета. Мы модели подаём на вход 701.2 и 702.1, ожидаем что она выдаст 700, а она выдаёт 2.18. Причём, очевидно, что она такое выдаст на что угодно. В приципе, логично что всё хуёво - слишком мало нейронов и итераций.

Добавим в середину слой из четырёх нейронов и посмотим не улучшится ли предсказательная сила модели. Заодно и увеличим количество итераций (циклов), которые обучается модель

In [41]:
# создаём модель заново, но +4 нейрона в середину

model = Sequential()
model.add(Dense(2, input_dim=2, kernel_initializer='normal', activation='relu'))
model.add(Dense(4, kernel_initializer='normal', activation='relu'))
model.add(Dense(1, kernel_initializer='normal'))
model.compile(loss='mean_squared_error', optimizer='adam')

In [42]:
# опять тренируем новую модель

model.fit(x_train, y_train, epochs=400)

Epoch 1/400
700/700 [==============================] - 0s 560us/step - loss: 162935.0984
Epoch 2/400
700/700 [==============================] - 0s 40us/step - loss: 162721.4978
Epoch 3/400
700/700 [==============================] - 0s 47us/step - loss: 162049.1255
Epoch 4/400
700/700 [==============================] - 0s 53us/step - loss: 160459.4539
Epoch 5/400
700/700 [==============================] - 0s 39us/step - loss: 157381.9571
Epoch 6/400
700/700 [==============================] - 0s 56us/step - loss: 152098.3003
Epoch 7/400
700/700 [==============================] - 0s 46us/step - loss: 144201

In [43]:
# посмотрим что модель предскажет

y_predicted = model.predict(x_test)

In [44]:
df = pd.DataFrame(x_test, columns=['x1_test', 'x2_test'])
df['y_predicted'] = y_predicted
df = df[['y_predicted', 'x1_test', 'x2_test']]
df.head()

y_predicted  x1_test  x2_test
0   699.524597    701.2    702.1
1   700.522705    702.2    703.1
2   701.520752    703.2    704.1
3   702.518738    704.2    705.1
4   703.516846    705.2    706.1

Ну уже поточнее. По мне так модель всё предсказывает хорошо и корреляция там очевидно предельно близка к 1

In [47]:
# более полно, если интересно

df

y_predicted  x1_test  x2_test
0     699.524597    701.2    702.1
1     700.522705    702.2    703.1
2     701.520752    703.2    704.1
3     702.518738    704.2    705.1
4     703.516846    705.2    706.1
5     704.514954    706.2    707.1
6     705.512878    707.2    708.1
7     706.510986    708.2    709.1
8     707.508972    709.2    710.1
9     708.507019    710.2    711.1
10    709.505066    711.2    712.1
11    710.503174    712.2    713.1
12    711.501221    713.2    714.1
13    712.499207    714.2    715.1
14    713.497314    715.2    716.1
15    714.495300    716.2    717.1
16    715.493347    717.2    718.1
17    716.491455    718.2    719.1
18    717.489441    719.2    720.1
19    718.487488    720.2    721.1
20    719.485535    721.2    722.1
21    720.483643    722.2    723.1
22    721.481567    723.2    724.1
23    722.479675    724.2    725.1
24    723.477661    725.2    726.1
25    724.475769    726.2    727.1
26    725.473816    727.2    728.1
27    726.471863    728.2    729.1
28    727.469910    729.2    730.1
29    728.467896    730.2    731.1
..           ...      ...      ...
270   968.996338    971.2    972.1
271   969.994446    972.2    973.1
272   970.992371    973.2    974.1
273   971.990479    974.2    975.1
274   972.988586    975.2    976.1
275   973.986572    976.2    977.1
276   974.984619    977.2    978.1
277   975.982666    978.2    979.1
278   976.980652    979.2    980.1
279   977.978699    980.2    981.1
280   978.976807    981.2    982.1
281   979.974854    982.2    983.1
282   980.972839    983.2    984.1
283   981.970947    984.2    985.1
284   982.968933    985.2    986.1
285   983.966980    986.2    987.1
286   984.965088    987.2    988.1
287   985.963074    988.2    989.1
288   986.961121    989.2    990.1
289   987.959229    990.2    991.1
290   988.957275    991.2    992.1
291   989.955261    992.2    993.1
292   990.953308    993.2    994.1
293   991.951416    994.2    995.1
294   992.949341    995.2    996.1
295   993.947449    996.2    997.1
296   994.945496    997.2    998.1
297   995.943542    998.2    999.1
298   996.941528    999.2   1000.1
299   997.939575   1000.2   1001.1

[300 rows x 3 columns]